crates/hir_ty/src/diagnostics/decl_check.rs

```diff
@@ -0,0 +1,821 @@
+//! Provides validators for the item declarations.
+//!
+//! This includes the following items:
+//!
+//! - variable bindings (e.g. `let x = foo();`)
+//! - struct fields (e.g. `struct Foo { field: u8 }`)
+//! - enum variants (e.g. `enum Foo { Variant { field: u8 } }`)
+//! - function/method arguments (e.g. `fn foo(arg: u8)`)
+//! - constants (e.g. `const FOO: u8 = 10;`)
+//! - static items (e.g. `static FOO: u8 = 10;`)
+//! - match arm bindings (e.g. `foo @ Some(_)`)
+
+mod str_helpers;
+
+use hir_def::{
+    adt::VariantData,
+    expr::{Pat, PatId},
+    src::HasSource,
+    AdtId, ConstId, EnumId, FunctionId, Lookup, ModuleDefId, StaticId, StructId,
+};
+use hir_expand::{
+    diagnostics::DiagnosticSink,
+    name::{AsName, Name},
+};
+use syntax::{
+    ast::{self, NameOwner},
+    AstNode, AstPtr,
+};
+
+use crate::{
+    db::HirDatabase,
+    diagnostics::{decl_check::str_helpers::*, CaseType, IncorrectCase},
+};
+
+pub(super) struct DeclValidator<'a, 'b: 'a> {
```

Comment:
This doesn't handle `allow()`  attributes, right?  I think this is fine for an experimental diagnostics, but we'll get a lot of reports for this nontheless :)

Reply:
Well, we can selectively disable diagnostics. Fellows that aren't interested in this feature can just add a `"incorrect-ident-case"` into the list of disabled diagnostics.

Isn't it enough? I think, parsing the crate attributes (especially given the fact that there are at least three different ones for this matter, and they may appear in different places) may become a burden.

Reply:
Well, it's a burden we'll have to carry anyway, at the end :-) It probably doesn't make to handle this in this PR, as it paves quite a bit of path already